packages/abi-coder/src/fragments/param-type.ts

```diff
@@ -0,0 +1,175 @@
+import { Logger } from '@ethersproject/logger';
+import { defineReadOnly } from '@ethersproject/properties';
+
+import { arrayRegEx, enumRegEx, structRegEx } from '../abi-coder';
+
+const version = 'abi/5.6.4';
+export const abiLogger = new Logger(version);
+
+export interface JsonFragmentType {
+  readonly name?: string;
+  readonly indexed?: boolean;
+  readonly type: string;
+  readonly components?: ReadonlyArray<JsonFragmentType>;
+  readonly typeArguments?: ReadonlyArray<JsonFragmentType>;
+}
+
+const CONSTRUCTOR_GUARD = {};
+
+function populate(object: ParamType, params: { type?: string } & ParamTypeProps) {
+  Object.keys(params).forEach((key) => {
+    const paramTypeKey = key as keyof ParamTypeProps;
+    const value = params[paramTypeKey];
+
+    defineReadOnly(object, paramTypeKey, value);
+  });
+}
+
+const paramTypeArray = /^(.*)\[([0-9]*)\]$/;
+
+export interface ParamTypeProps {
+  // The local name of the parameter (of null if unbound)
+  readonly name?: string;
+
+  // The fully qualified type (e.g. "address", "tuple(address)", "uint256[3][]"
+  readonly type: string;
+
+  // The base type (e.g. "address", "tuple", "array")
+  readonly baseType?: string;
```

Comment:
I think we have the opportunity now to transform BaseType into an Enum. Maybe we should open a issue for this.